include/fiction/algorithms/simulation/sidb/random_layout_generator.hpp

```diff
@@ -0,0 +1,200 @@
+                // coordinate is constructed [x_coordinate, y_coordinate]
+                const auto random_coordinate = typename Lyt::coordinate({random_x_coordinate, random_y_coordinate});
+
+                bool constraint_violation_positive_sidbs = false;
+
+                if (parameter.prevent_positive_charges)
+                {
+                    // it checks if new coordinate is not closer than 2 cells (Euclidean distance).
+                    lyt.foreach_cell(
+                        [this, &lyt, &random_coordinate, &constraint_violation_positive_sidbs](const auto& c1)
+                        {
+                            if (euclidean_distance<Lyt>(lyt, c1, random_coordinate) < parameter.minimal_spacing)
+                            {
+                                constraint_violation_positive_sidbs = true;
+                            }
+                        });
+                }
+
+                if (!constraint_violation_positive_sidbs)
+                {
+                    lyt.assign_cell_type(random_coordinate, Lyt::cell_type::NORMAL);
+                }
+                attempt_counter += 1;
+            }
+
+            // it checks if new-found layout is identical to an already found layout (all_layouts).
+            uint64_t identical_layout_counter = 0;
+            for (const auto& old_lyt : previous_layouts)
+            {
+                old_lyt.foreach_cell(
+                    [&identical_layout_counter, lyt](const auto& cell_old) mutable
+                    {
+                        lyt.foreach_cell(
+                            [&identical_layout_counter, &cell_old](const auto& cell_new) mutable
+                            {
+                                if (cell_new == cell_old)
+                                {
+                                    identical_layout_counter += 1;
+                                }
+                            });
+                    });
+            }
+
+            // if all SiDBs are placed and the new layout is not a duplication, the layout is written to the
+            // std::ostream.
+            if (lyt.num_cells() == parameter.number_placed_sidbs && identical_layout_counter == 0)
+            {
+                layout = lyt;
+                write_sqd_layout(lyt, os);
+                successful_generation = true;
+            }
+        }
+    }
+
+    Lyt get_layout() const
+    {
+        return layout;
+    }
+
+  private:
+    Lyt                       layout;
+    random_layout_params<Lyt> parameter;
+    std::ostream&             os;
+    std::vector<Lyt>          previous_layouts;
+};
+}  // namespace detail
+
+/**
+ * This algorithm generates a layout with randomly distributed SiDBs.
+ *
+ * @tparam Lyt Cell-level with fiction coordinates.
+ * @param params All parameters needed for the generation of random layouts.
+ * @param os The output stream to write into.
+ * @param all_layouts Previous generated layouts to avoid duplication.
```

Comment:
This parameter doesn't exist. 

Also, is it really necessary to check for duplication? Isn't the chance vanishingly small?

How about a compromise? Have this function to generate single random layouts. Create another one to generate multiple layouts (parameterized by a number) that calls this function and keeps an `unordered_set` of the layouts. It can then quickly check whether it has a duplicate. Sounds good?

Reply:
That's a great idea! However, this would require defining an equal operator for the cell_level_layout. Please check if the updated version (after I have re-request the review) is ok for you.